### ISSUE TITLE
Unable to Run DOS Game Amika

### ISSUE THREAD

**Issue by nightson** [joncampbell123/dosbox-x#2338]
**Describe the bug**
The game doesn't start properly. DOSBox-X only shows a black screen.

**Debug log**
COMMAND.COM env size:             720 bytes
COMMAND.COM environment block:    0x0701 sz=0x002d
COMMAND.COM main body (PSP):      0x072f sz=0x009a
COMMAND.COM stack:                0x0749
pixratio 1.350, dw false, dh false
Aspect ratio: 720 x 540  xToY=1.333 yToX=0.750
Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (-1.000 x -1.000 mm) (-0.039 x -0.039 in) (-1.000 x -1.000 DPI)
Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (-1.000 x -1.000 mm) (-0.039 x -0.039 in) (-1.000 x -1.000 DPI)
Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (-1.000 x -1.000 mm) (-0.039 x -0.039 in) (-1.000 x -1.000 DPI)
Aspect ratio: 720 x 540  xToY=1.333 yToX=0.750
Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (-1.000 x -1.000 mm) (-0.039 x -0.039 in) (-1.000 x -1.000 DPI)
  36063977 ERROR BIOS:INT15:Unknown call ax=BFDE
  36064059 ERROR BIOS:INT15:Unknown call ax=BFDE
DOSBox-X has switched to max cycles, because of the setting: cycles=auto.
If the game runs too fast, try a fixed cycles amount in DOSBox-X's options.
  36074793 ERROR BIOS:INT15:Unknown call ax=BF01
  38543604 ERROR MOUSE:Unhandled videomode 69 on reset
pixratio 1.000, dw false, dh false
Aspect ratio: 640 x 480  xToY=1.333 yToX=0.750
Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (-1.000 x -1.000 mm) (-0.039 x -0.039 in) (-1.000 x -1.000 DPI)
Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (-1.000 x -1.000 mm) (-0.039 x -0.039 in) (-1.000 x -1.000 DPI)
Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (-1.000 x -1.000 mm) (-0.039 x -0.039 in) (-1.000 x -1.000 DPI)
Aspect ratio: 640 x 480  xToY=1.333 yToX=0.750
Screen report: Method 'Win98base' (1920.000 x 1080.000 pixels) at (0.000 x 0.000) (-1.000 x -1.000 mm) (-0.039 x -0.039 in) (-1.000 x -1.000 DPI)

**Environment:**
 - Windows 7 Ultimate SP1 64-Bit
 - DOSBox-X 0.83.11
 - Default Config

**Additional context**
The game runs just fine under real DOS. If you boot a real DOS 6.22 from an image inside DOSBox-X the game will also run without any special configuration.

**Comment by rderooy** [comment]
I cannot find this game on Mobygames.

Can you try running "loadfix -a" and then starting the game?

**Comment by nightson** [comment]
> 
> 
> I cannot find this game on Mobygames.
> 
> Can you try running "loadfix -a" and then starting the game?

Didn't work. I still got the same error messages.

It's a Chinese game and like most of them it's not on Mobygames yet. You can find it on some abandonware sites though. Using the Chinese name 精武戰警 might be easier :)

**Comment by rderooy** [comment]
Not the easiest to find. Here is a RAR file which contains windows dosbox plus a IMG file that boots DOS and then runs the game.
http://m.kuaihou.com/game/55042.html
Chrome said there was something suspicious about the archive, and immediately quarantined it. I just deleted the included Windows binaries and kept the IMG file.

From what I gather the game would have been on a CD with sound tracks that are missing from this copy?

In any case, from this copy, I can indeed confirm the black screen when running it in DOSBox-X directly. Running it as mentioned by first booting DOS does work. The same thing also happens with vanilla DOSBox.

**Comment by joncampbell123** [comment]
Can you try adding this to your dosbox.conf?

[speaker]
pcspeaker clock gate enable at startup=true

It may have the same problem as test case "Baoxiao Sanguozhi" submitted a year or so ago, using the PC speaker PIT timer output for timed delay loops, but assumes the PC speaker clock gate is on instead of turning it on to make it work.

(will test the IMG file now)

**Comment by rderooy** [comment]
@joncampbell123 

Setting the pcspeaker option made no difference here.

**Comment by joncampbell123** [comment]
@rderooy Right, same result here. I'm seeing it sit in DOSBox's keyboard input waiting loop here.
Booting the image works but the mouse doesn't move at the menu.

**Comment by rderooy** [comment]
@joncampbell123 I noticed the same thing when booting the IMG. But when I boot the IMG with vanilla dosbox, the mouse does work. It's movement area is being constraint to just over the buttons.

**Comment by joncampbell123** [comment]
Also, according to the debugger, this game is making unnecessary calls to INT 21h AX=FF00h. It doesn't appear to use the result, though.

http://www.ctyme.com/intr/rb-4059.htm

**Comment by rderooy** [comment]
TopWare Network OS? never heard of it, and cannot find anything about it online either apart from it's interrupt list.

**Comment by joncampbell123** [comment]
It seems to have something to do with access to 0xE0000000 (the linear SVGA framebuffer), and page faults. It just loops through two calls to bank switch with INT 10h not advancing. If I step through with the debugger it eventually goes into a recursive page fault for it. The latest DOSBox SVN on my hard drive also goes into an infinite page fault recursion loop as well and crashes. Strange.

**Comment by joncampbell123** [comment]
Uh... I think this explains why the mouse isn't working when booting the image in DOSBox-X.

I never enabled serial mouse emulation.
![guest os_000](https://user-images.githubusercontent.com/6245486/110254707-1400cd80-7f45-11eb-8fa4-87cfb01cc2cc.png)

**Comment by joncampbell123** [comment]
Disabling the COM ports in dosbox.conf and... it doesn't like DOSBox-X PS/2 mouse emulation?

![guest os_001](https://user-images.githubusercontent.com/6245486/110254761-5a562c80-7f45-11eb-842b-e41c4f4b090d.png)

**Comment by joncampbell123** [comment]
MSCMOUSE is rejecting DOSBox-X's PS/2 mouse because INT 15h AX=C201h (POINTING DEVICE RESET) is returning error code 0x02 invalid input. Looking at src/ints/bios.cpp, a recent pull request made function 01h (RESET) a fall through to 05h (INITIALIZE). The fall through sets BX=00AAh which the INITIALIZE call rejects because BH is not 3 or 4 (packet size), returning code 2.

**Comment by joncampbell123** [comment]
Fixing INT 15h AX=C201h to return success fixes the issue, makes MSCMOUSE happy.
![guest os_002](https://user-images.githubusercontent.com/6245486/110255056-c5ecc980-7f46-11eb-8bd8-01c08054e596.png)

**Comment by joncampbell123** [comment]
So when run in DOSBox SVN and DOSBox-X's native DOS environment, AMIKA gets stuck at a REP MOVSB instruction trying to resolve the page fault to call INT 10h AX=4F05h to bank switch (why there's a page fault to E0000000h I'm not sure).

**Comment by joncampbell123** [comment]
According to the debugger, this is one of those games (like the quirky Demoscene stuff I've tested against) that would rather use the bank switched SVGA modes and page fault tricks to fake a linear framebuffer (mode 0x101 INT 10h AX=4F02h) than just ask for the linear framebuffer.

To be fair, it's the same trick Windows 95 uses to simplify it's GDI and graphics support: Write the code around a flat linear framebuffer, and if the ISA card is the old bank switching type, use page faults to fake a linear framebuffer.

**Comment by joncampbell123** [comment]
Uh, wait, actually the debugger is showing it stuck in a loop polling I/O port 006h? First while bit 0 is set, then bit 0 is clear.

Scrolling up, it loads DX from a 16-bit unsigned integer in memory, adds 6.

What's significant about that you ask?

Well, the loop suggests it's supposed to be polling port 3DAh, and the value loaded is supposed to the the CRTC index I/O port 3D4h / 3B4h.

It's zero for some reason.

Either the BIOS data area is not providing it or it expects it from the VESA BIOS extensions somehow.

**Comment by joncampbell123** [comment]
Still not sure what's going on from the DOSBox native environment, but booting the disk image should work now with a working mouse if you compile the latest commit.

**Comment by ualaa1** [comment]
I did a quick test of the Korean CD version (I can provide links if you want) on 0.83.11 and it seems to run fine. Maybe the problem is that the Chinese img file was improperly ripped from the lost Chinese CD version?

**Comment by nightson** [comment]
> I did a quick test of the Korean CD version (I can provide links if you want) on 0.83.11 and it seems to run fine. Maybe the problem is that the Chinese img file was improperly ripped from the lost Chinese CD version?

Nah. The same issue still exists even if you use original CDs. Also I don't think improperly ripped images would run in native DOS environment without any issue.

**Comment by nightson** [comment]
> Still not sure what's going on from the DOSBox native environment, but booting the disk image should work now with a working mouse if you compile the latest commit.

It would be better if we could run it in native DOSBox environment. Should I report this to the vanilla DOSBox developers?

By the way, what knowledge does it require for one to know how to debug a game when something went wrong? Is there some good website or book to start with?

**Comment by ualaa1** [comment]
P

> > I did a quick test of the Korean CD version (I can provide links if you want) on 0.83.11 and it seems to run fine. Maybe the problem is that the Chinese img file was improperly ripped from the lost Chinese CD version?
> 
> Nah. The same issue still exists even if you use original CDs. Also I don't think improperly ripped images would run in native DOS environment without any issue.

Proof that it runs fine (at least for me) on default settings. No mouse problem whatsoever.:
![amika_002](https://user-images.githubusercontent.com/68984176/110444701-304e5880-8072-11eb-8bbc-aca272e48ae3.png)
![amika_004](https://user-images.githubusercontent.com/68984176/110444713-35130c80-8072-11eb-9949-d1e873ef4eea.png)
![amika_005](https://user-images.githubusercontent.com/68984176/110444726-393f2a00-8072-11eb-85c8-8bd083abd63f.png)

**Comment by joncampbell123** [comment]
It requires patience with the debugger interface, breakpoint mechanism (BPINT) and the Ralph Brown Interrupt List.

The game seems to have problems resolving the page fault it makes for itself in faking a linear framebuffer from bank switching. You can tell when it calls INT 10h AX=4F02h to set the VBE mode, it uses mode 0x101 (bank switching 640x480), not mode 0x4101 (linear framebuffer 640x480). The page fault handler calls INT 10h AX=4F05h to change the active bank. Notice tracing from that it runs back through to an IRET which comes back to a REP MOVSB that caused the fault in the first place. Under the DOSBox-X DOS kernel that seems to cause a fault again.

http://hackipedia.org/browse.cgi/Computer/Platform/PC%2c%20IBM%20compatible/BIOS/Video/VESA%20BIOS%20Extensions/VESA%20BIOS%20extensions%20%281998%2d09%2d16%29%20v3%2e0%2epdf
